### PR TITLE
Devkelley/standardize extension model

### PR DESCRIPTION
**Pull request by devkelley**
This PR refactors the extension library and converts the naming to 'Module'. In addition:
- Makes each module a self contained library with its own Cargo.toml file. (core/module/<Module>/*)
- Adds a Grpc_Server object that assists with adding Modules and their interceptors to a builder and then building all the modules into a single server.
- Moves the logic for adding modules with the feature flags to the main.rs file.
- Moves 'load_settings' into the utils.rs file.

### CODE REVIEW

File: core/common/src/grpc_module.rs [hunk 1]
```diff
@@ -4,9 +4,9 @@
 
 use tonic::transport::server::RoutesBuilder;
 
-/// Trait that must be implemented for an extension to add a grpc service to the hosted server.
-/// Note: This trait may be renamed in the future.
-pub trait GrpcExtension {
+/// Trait that must be implemented for a module to add one or more grpc services to the hosted
```
Comment (ashbeitz) [inline core/common/src/grpc_module.rs:7]: Should this description also mention layers/middleware?

File: core/common/src/grpc_server.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    ///
+    /// # Arguments
+    /// * `address` - The address the server will be hosted on.
+    pub fn new(address: SocketAddr) -> Self {
+        GrpcServer { address, modules: RoutesBuilder::default(), middleware: ServiceBuilder::new() }
+    }
+}
+
+impl<L> GrpcServer<L> {
+    /// Adds a module (collection of grpc services and interceptors) for the server to host.
```
Comment (ashbeitz) [inline core/common/src/grpc_server.rs:32]: ... and grpc interceptors ...

File: core/invehicle-digital-twin/src/main.rs [hunk 3]
```diff
@@ -61,6 +77,59 @@ async fn register_invehicle_digital_twin_service_with_chariott(
     Ok(())
 }
 
+/// Builds the enabled modules for the grpc server and starts the server.
+///
+/// # Arguments
+/// * `addr` - The address the server will be hosted on.
+/// * `base_service` - The core service that will be hosted.
+///
+/// # How to add an Module to this method:
```
Comment (ashbeitz) [inline core/invehicle-digital-twin/src/main.rs:86]: .. a Module ..

File: core/invehicle-digital-twin/src/main.rs [hunk 4]
```diff
@@ -61,6 +77,59 @@ async fn register_invehicle_digital_twin_service_with_chariott(
...
+/// 1. Add a block of code with the appropriate cfg feature flag.
+/// 2. Create the `GrpcModule` object within the block - if applicable.
+/// 3. Create the `GrpcInterceptorLayer` object(s) within the block - if applicable.
+/// 4. Add the interceptors to the middleware stack with `.layer()`.
```
Comment (ashbeitz) [inline core/invehicle-digital-twin/src/main.rs:90]: ... the grpc interceptors ..

File: core/invehicle-digital-twin/src/main.rs [hunk 5]
```diff
@@ -61,6 +77,59 @@ async fn register_invehicle_digital_twin_service_with_chariott(
...
+        + 'static,
+    S::Future: Send + 'static,
+{
+    let mut server: GrpcServer<Identity> = GrpcServer::new(addr);
+
+    #[cfg(feature = "managed_subscribe")]
+    // (1) Adds the Managed Subscribe module to the service.
+    let server = {
+        // (2) Initialize the Managed Subscribe module, which implements GrpcModule.
+        let managed_subscribe_ext = ManagedSubscribeExt::new();
```
Comment (ashbeitz) [inline core/invehicle-digital-twin/src/main.rs:112]: Perhaps rename it to "managed_subscribe_module".

Comment (ashbeitz) [reply core/invehicle-digital-twin/src/main.rs:112]: And rename the associated struct name.

File: core/module/managed_subscribe/src/managed_subscribe_ext.rs [hunk 6]
```diff
@@ -148,7 +152,7 @@ impl ManagedSubscribeExt {
     }
 }
 
-impl GrpcExtension for ManagedSubscribeExt {
+impl GrpcModule for ManagedSubscribeExt {
```
Comment (ashbeitz) [inline core/module/managed_subscribe/src/managed_subscribe_ext.rs:155]: Rename ManagedSubscribeExt, as we are moving way from using "extension".

**Review by ashbeitz — COMMENTED** [review]
Please take a look at my comments.

File: core/common/src/grpc_module.rs (hunk 1 above)
Comment (devkelley) [reply core/common/src/grpc_module.rs:7]: Updated

File: core/common/src/grpc_server.rs (hunk 2 above)
Comment (devkelley) [reply core/common/src/grpc_server.rs]: changed

File: core/invehicle-digital-twin/src/main.rs (hunk 3 above)
Comment (devkelley) [reply core/invehicle-digital-twin/src/main.rs]: changed

File: core/invehicle-digital-twin/src/main.rs (hunk 4 above)
Comment (devkelley) [reply core/invehicle-digital-twin/src/main.rs]: changed

File: core/invehicle-digital-twin/src/main.rs (hunk 5 above)
Comment (devkelley) [reply core/invehicle-digital-twin/src/main.rs]: updated here and elsewhere!

File: core/module/managed_subscribe/src/managed_subscribe_ext.rs (hunk 6 above)
Comment (devkelley) [reply core/module/managed_subscribe/src/managed_subscribe_ext.rs]: Updated here and elsewhere

**Review by ashbeitz — APPROVED** [review]
Approved.

**Pull request merged**